src/services/trackingPlan.ts

```diff
@@ -1,88 +1,83 @@
+import { HTTP_STATUS_CODES } from '../v0/util/constant';
 
 export class TrackingPlanservice {
-  public static async validateTrackingPlan(events, requestSize, reqParams) {
-    const requestStartTime = new Date();
+  public static async validate(events, requestSize, reqParams) {
+    const startTime = new Date();
     const respList: any[] = [];
-    const metaTags = events[0].metadata ? getMetadata(events[0].metadata) : {};
+    const metaTags = events.length && events[0].metadata ? getMetadata(events[0].metadata) : {};
+    const tpTags =
+      events.length && events[0].metadata ? getTrackingPlanMetadata(events[0].metadata) : {};
     let ctxStatusCode = 200;
 
-    for (let i = 0; i < events.length; i++) {
-      const event = events[i];
-      const eventStartTime = new Date();
+    for (let event of events) {
+      let eventValidationResponse: any;
+      let exceptionOccured = false;
+      let eventStartTime = new Date();
+
       try {
-        const parsedEvent = event;
-        parsedEvent.request = { query: reqParams };
-        const hv = await eventValidator.handleValidation(parsedEvent);
-        if (hv['dropEvent']) {
-          respList.push({
-            output: event.message,
-            metadata: event.metadata,
-            statusCode: 400,
-            validationErrors: hv['validationErrors'],
-            error: JSON.stringify(constructValidationErrors(hv['validationErrors'])),
-          });
-          stats.counter('tp_violation_type', 1, {
-            violationType: hv['violationType'],
-            ...metaTags,
-          });
-        } else {
-          respList.push({
-            output: event.message,
-            metadata: event.metadata,
-            statusCode: 200,
-            validationErrors: hv['validationErrors'],
-            error: JSON.stringify(constructValidationErrors(hv['validationErrors'])),
-          });
-          stats.counter('tp_propagated_events', 1, {
-            ...metaTags,
-          });
-        }
+        event.request = { query: reqParams };
+        const validatedEvent = await eventValidator.handleValidation(event);
+        eventValidationResponse = {
+          output: event.message,
+          metadata: event.metadata,
+          statusCode: validatedEvent['dropEvent']
+            ? HTTP_STATUS_CODES.BAD_REQUEST
+            : HTTP_STATUS_CODES.OK,
+          validationErrors: validatedEvent['validationErrors'],
+          error: JSON.stringify(constructValidationErrors(validatedEvent['validationErrors'])),
+        };
       } catch (error) {
-        const errMessage = `Error occurred while validating : ${error}`;
-        logger.error(errMessage);
-        let status = 200;
+        logger.debug(
+          `Error occurred while validating event`,
+          'event',
+          event?.eventName,
+          'tp',
+          event?.trackingPlanId,
+          'tpVersion',
+          event?.trackingPlanVersion,
+          'error',
+          error,
+        );
```

Comment:
```suggestion
        logger.debug(
          `Error occurred while validating event ${event.message?.event} using tracking plan ${tpTags.trackingPlanId}: ${error.message}`,
        );
```

Reply:
I think I forgot to put `metadata` as the access key on the main object but isn't it better to put `metadata` as the key there ?

Reply:
change looks fine now.